### ISSUE TITLE
Pretty Print for Dataset Objects

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1816]
Beginners have trouble working with our `Dataset` objects in part because they're opaque. Here's how a `Dataset` object prints on the REPL:

```
>>> dc.data.NumpyDataset(np.random.rand(5,5))
<deepchem.data.datasets.NumpyDataset at 0x1a454f3a20>
```

I propose that we add pretty printers for dataset objects. Here's my proposed simple pretty printer:

```
  def __repr__(self):
      return "<%s X.shape: %s, y.shape: %s, w.shape: %s, ids: %s, task_names: %s>" % (self.__class__.__name__, str(self.X.shape), str(self.y.sh
ape), str(self.w.shape), self.ids.__repr__(), self.get_task_names())
 
  def __str__(self):                                                   
      return "<%s X.shape: %s, y.shape: %s, w.shape: %s, ids: %s, task_names: %s>" % (self.__class__.__name__, str(self.X.shape), str(self.y.sh
ape), str(self.w.shape), self.ids.__repr__(), self.get_task_names())
```

It prints shapes for the features, but prints the id arrays outright. This makes things like splitters work very nicely. Here's the output of a simple scaffold splitting script:

```
Original set of molecules
['C1=CC2=C(C=C1)C1=CC=CC=C21', 'O=C1C=CC(=O)C2=C1OC=CO2', 'C1=C[N]C=C1', 'C1=CC=CC=C[C+]1', 'C1=[C]NC=C1', 'N[C@@H](C)C(=O)O', 'N[C@H](C)C(=O)O', 'CC', 'O=C=O', 'C#N', 'CCN(CC)CC', 'CC(=O)O', 'C1CCCCC1', 'c1ccccc1']
Training set
<NumpyDataset X.shape: (11,), y.shape: (11, 1), w.shape: (11, 1), ids: array(['N[C@@H](C)C(=O)O', 'N[C@H](C)C(=O)O', 'CC', 'O=C=O', 'C#N',
       'CCN(CC)CC', 'CC(=O)O', 'c1ccccc1', 'C1CCCCC1', 'C1=[C]NC=C1',
       'C1=CC=CC=C[C+]1'], dtype=object), task_names: [0]>
Valid set
<NumpyDataset X.shape: (1,), y.shape: (1, 1), w.shape: (1, 1), ids: array(['C1=C[N]C=C1'], dtype=object), task_names: [0]>
Test set
<NumpyDataset X.shape: (2,), y.shape: (2, 1), w.shape: (2, 1), ids: array(['O=C1C=CC(=O)C2=C1OC=CO2', 'C1=CC2=C(C=C1)C1=CC=CC=C21'],
      dtype=object), task_names: [0]>
```

Notice how it's very clear which compounds are in which split with the pretty print enabled. The downside of this printer is that printing a large dataset can get unwieldy with all the ids printed.

I'd like some feedback on this proposed pretty print design. Are there ways we can make this print more robust or configurable? Are there other things we might want to add to have a well featured pretty-printer? 

CC @peastman @vsomnath

**Comment by peastman** [comment]
> The downside of this printer is that printing a large dataset can get unwieldy with all the ids printed.

For the ids and task names, how about having it print the full list if it's less than some reasonable length, but only the size if it's longer than that?

**Comment by rbharath** [comment]
This was implemented in #1844.
